### DISCUSSION TITLE
Confluence Integration for Knowledge Sources - Design Discussion

### DISCUSSION THREAD

**Discussion by gabriel-ecegi** [open-webui/open-webui#7888] [Ideas]
# Confluence Integration for Knowledge Sources - Design Discussion

## Overview
Building on the discussion in #5872, this proposal outlines a comprehensive approach for integrating Confluence and other knowledge sources, including scheduling and admin management capabilities.

### 1. Connector Architecture
```
knowledge_sync/
  main.py #syncing connectors logic
  connectors/
    confluence/
      sync/         # Core sync logic
      pages/        # Page handling
      spaces/       # Space handling
      config.py     # Connector config
```

The connector remains modular with separated concerns.

### 2. Scheduled Sync Jobs
```python
@asynccontextmanager
async def lifespan(app: FastAPI):
    scheduler = AsyncIOScheduler()
    scheduler.add_job(sync_job, 'interval', hours=24, id='knowledge_sync')
    scheduler.start()
    yield
```

- **Periodic Syncs**: Automated daily syncs keep content fresh
- **Initial Sync**: Optional immediate sync on startup
- **Configurable Schedule**: Admins can adjust sync frequency
- **Job Management**: Monitor job status and history

### 3. Admin Interface
New "Data Sources" admin section would allow:

- Source configuration and credentials
- Sync schedule management  
- Status monitoring
- Permission mapping configuration
- Custom source registration

### 4. Admin-Generated Sources
Enable Admins to:

- Register personal data sources
- Upload custom documents
- Configure source-specific settings
- Monitor their source sync status

## Discussion Points

1. **Admin UI Organization**: Where should data source management live in the admin interface? Options:
   - New top-level "Data Sources" section
   - Under existing "Documents" section
   - Separate "Knowledge Management" area

2. **Permission Inheritance**: How should source permissions map to WebUI permissions?
   - Direct mapping vs simplified model
   - Handling conflicts between sources
   - Role-based access control

## Next Steps
- Decide admin UI
- Define admin source registration flow
- Implement job scheduling framework
- Create source management API endpoints

Would love feedback on:
1. Admin UI placement and organization
2. Admin source registration workflow
3. Permission model standardization
4. Sync scheduling configuration

**Comment by ColWhackaMole** [comment]
this would be amazing if made a reality

**Comment by arty-hlr** [comment]
Absolutely, worth having for sure!

**Comment by phlashworth** [comment]
You may be able to draw some inspiration from [Onxy](https://github.com/onyx-dot-app/onyx/) (previously Danswer) as it has a [confluence connector](https://github.com/onyx-dot-app/onyx/tree/main/backend/onyx/connectors/confluence).

**Reply by phlashworth** [reply]
Looks to be MIT licensed with the exception of some ee parts of the repo.

**Comment by yuhongsun96** [comment]
Have you all considered using Onyx instead 🥺 (cocreator/maintainer here)? Why rebuild our software 😭

We have a pretty sick chat UI too

**Reply by phlashworth** [reply]
I can confirm that having taken a look, I'm very impressed with the Onyx project! It does seem that both Onyx and Open WebUI both have their strengths and weaknesses but in this case, connectors to external data sources seems to be one of Onyx's strengths!

**Comment by elabbarw** [comment]
Why not simplify things? Instead of introducing a centralized admin feature—especially when managing user permissions and access to various spaces can get tricky—allow users to add documents to their existing knowledge base. They would do this using CQL along with an access token that's requested and stored against their account, or used every time they want to sync so you don't need to have the added complications of having a secret vault as well.

When users want to add content to their KB, they would follow this flow:
- Click "+" then "Sync Confluence CQL".
- Enter confluence cloud/on-prem server url along with access token that they have to generate.
- OWUI pulls the CQL results, retrieving all documents as Markdown(already provided by API), along with their last updated timestamps.
- The method for adding text content is triggered, which integrates the embedding and saving processes.

For updates, the user simply reinitiates the process, re-enters their token, and any existing documents from the CQL results that match will be updated by checking the last updated timestamp from confluence against the timestamp of when the doc was created. 


Let's start small 😄

**Comment by RomainNeup** [comment]
If anyone is looking for a tool for it, I built 2 tools : [confluence search](https://openwebui.com/t/romainneup/confluence_search) and [confluence get page by id](https://openwebui.com/t/romainneup/confluence_page).
Source code is available here : https://github.com/RomainNeup/open-webui-utilities
Hope it will fits your needs (and if not don't hesitate to open an issue on Github)

**Reply by enving** [reply]
Hey Romain, thanks very much. I just had a look to your tools..do they have still relevance from your perspective now that there might be mcp server for confluence https://github.com/sooperset/mcp-atlassian and openwebui integrating mcp https://github.com/open-webui/open-webui/discussions/7363 ? what do you think of it? :) Cheers

**Reply by RomainNeup** [reply]
Hello, yep it does for multiple reasons. First one is that OWUI is not supporting MCP yet (at least the official version). Second one is that you need to run a MCP server which is a bit more complex/ressource demanding than just running a tool directly in OWUI.
But indeed I plan to create a MCP version of the tool (if I need feature not present in the actual one or maybe contribute the the MCP tool) when OWUI would support MCP officially
